### ISSUE TITLE
[Bug] Popup Dismiss passing an object back crashes unless invoked on main thread iOS only 

### ISSUE THREAD

**Issue by stephenhauck** [xamarin/XamarinCommunityToolkit#1619]
Dismiss will crash with a UI View error on  iOS, android seems to work fine
`Dismiss(result);`

### Steps to Reproduce
1.  Download my example app here https://github.com/stephenhauck/ZXingPopup
2. Change this code  `MainThread.BeginInvokeOnMainThread(() =>
                {
                    Dismiss(result);
                });`

To          `Dismiss(result);`

3.  Run app on iOS device

Running on latest stable everything ...

**Comment by stephenhauck** [comment]
When they cancel I pass back null and all is well ...

**Comment by ahoefling** [comment]
The title of this bug says it only works on the main UI thread for iOS. The popups are required to operate on the main UI thread, any exceptions caused because it is not operating on the main UI thread is an expected behavior. There is no trick or workaround to this as the popup operation will need to be marshaled from a background thread to the main UI thread. Blindly marshaling all threads to the main UI thread can cause issues with the thread pool which may slow the app down or cause other performance issues.

cc: @pictos

**Comment by stephenhauck** [comment]
ok but it's not our job to call "invoke on main thread" right ?
That should be handled in the "Dismiss" ?
When I "invoke on main thread" a blink occurs and it's just not right

**Comment by pictos** [comment]
@ahoefling this is a tricky one... Normally frameworks don't do the work to invoke things in the UI thread. Is the responsibility of the mobile dev to make sure the UI changes are happening in the UI thread. CollectionView had an issue because [of that](https://github.com/xamarin/Xamarin.Forms/issues/12080)

I would say that is not the job of the lib to force it to be on the UI thread because doing this in _best way possible_ will be very hard.

> but it's not our job to call "invoke on main thread" right?

I would say if you, as a mobile dev, are doing some work on the background thread is it your job to return to the main thread. Like

```csharp
void Do() {
Task.Run( ( ) => async {
    string result = await httpClient.GetString("SomeUrl").ConfigureAwait(false);
    label.Text = result;
}
```

If you try to do this on XF today, probably it will fail, because you are changing the `label.Text` in a background thread... So at the end of the day, the framework can't do this for devs.

**Comment by ahoefling** [comment]
I agree with @pictos evaluation, I don't believe this is a bug. The popup control is built tightly around the native implementations and it is expected to used from the main UI thread. Your application code can perform processing off the main thread, but when it is time to perform UI operations you will need to go back to the main UI thread.
